cmd/credentialUtil.go

```diff
@@ -257,35 +257,146 @@ type rawFromToInfo struct {
+const trustedSuffixesNameAAD = "trusted-aad-suffixes"
+const trustedSuffixesNameS3 = "trusted-s3-suffixes"
+const trustedSuffixesAAD = "*.core.windows.net;*.core.chinacloudapi.cn;*.core.cloudapi.de;*.core.usgovcloudapi.net"
+const trustedSuffixesS3 = "*.amazonaws.com;*.amazonaws.com.cn"
+
+// checkAuthSafeForTarget checks our "implicit" auth types (those that pick up creds from the environment
+// or a prior login) to make sure they are only being used in places where we know those auth types are safe.
+// This prevents, for example, us accidentally sending OAuth creds to some place they don't belong
+func checkAuthSafeForTarget(ct common.CredentialType, resource, extraSuffixesAAD, extraSuffixesS3 string) error {
+
+	getSuffixes := func(list string, extras string) []string {
+		extras = strings.Trim(extras, " ")
+		if extras != "" {
+			list += ";" + extras
+		}
+		return strings.Split(list, ";")
+	}
+
+	isResourceInSuffixList := func(suffixes []string) (string, bool) {
+		u, err := url.Parse(resource)
+		if err != nil {
+			return "<unparsable>", false
+		}
+		host := strings.ToLower(u.Host)
+
+		for _, s := range suffixes {
+			s = strings.Trim(s, " *") // trim *.foo to .foo
+			s = strings.ToLower(s)
+			if strings.HasSuffix(host, s) {
+				return host, true
+			}
+		}
+		return host, false
+	}
+
+	switch ct {
+	case common.ECredentialType.Unknown(),
+		common.ECredentialType.Anonymous():
+		// these auth types don't pick up anything from environment vars, so they are not the focus of this routine
+		return nil
+	case common.ECredentialType.OAuthToken(),
+		common.ECredentialType.SharedKey():
+		// these are Azure auth types, so make sure the resource is known to be in Azure
+		domainSuffixes := getSuffixes(trustedSuffixesAAD, extraSuffixesAAD)
+		if host, ok := isResourceInSuffixList(domainSuffixes); !ok {
+			return fmt.Errorf(
+				"azure authentication to %s is not enabled in AzCopy. To enable, view the documentation for "+
+					"the parameter --%s, by running 'AzCopy copy --help'. Then use that parameter in your command if necessary",
+				host, trustedSuffixesNameAAD)
+		}
+
+	case common.ECredentialType.S3AccessKey():
+		// make sure the resource is known to be in AWS
+		domainSuffixes := getSuffixes(trustedSuffixesS3, extraSuffixesS3)
+		if host, ok := isResourceInSuffixList(domainSuffixes); !ok {
+			return fmt.Errorf(
+				"s3 authentication to %s is not enabled in AzCopy. To enable, view the documentation for "+
+					"the parameter --%s, by running 'AzCopy copy --help'. Then use that parameter in your command if necessary",
+				host, trustedSuffixesNameS3)
+		}
+
+	default:
+		panic("unknown credential type")
+	}
+
+	return nil
+}
+
+func logAuthType(ct common.CredentialType, location common.Location, isSource bool) {
+	if location == common.ELocation.Unknown() {
```

Comment:
We may want to do this once per auth location, if not already done.

Reply:
It does it once per message, which amounts to the same thing.